### DISCUSSION TITLE
Full-Featured API to remotly control Pybricks firmware ?

### DISCUSSION THREAD

**Discussion by JeanRev** [pybricks/support#615] [General]
Hi there,

Last week I played with my son's Move Hub, with the idea of controlling it with an Xbox bluetooth controller.

After a few hours of effort, the result is quite satisfying I must say. Input lag is minimal, and the joystick allows really fine speed adjustment:
https://user-images.githubusercontent.com/8666989/150674999-1cc7f926-968e-45aa-96c1-c919b1ca11cd.mp4

Currently the code runs on a Raspberry 3B and uses the [node-powered up](https://github.com/nathankellenicki/node-poweredup) library to communicate with the hub. Node-poweredup allows to remotely control the official lego firmware by offering a very simple and intuitive API.

To make it short: I would have loved to be able to do the same thing with PyBricks. The idea of having a programmable firmware is really awesome, but I would also like to have a simple way to remotely control my creations. The best of both worlds so to speak :-) If I understood correctly, there is currently no such API for PyBricks. Sending command to the hub must be done with the somewhat low-level function getchar, which requires to implements some kind of communication protocol between the client and the hub.

Would it be worthwhile to enter a feature request for such client API? Is that too ambitious ? Or would it not really fit with Pybricks' philosophy?

(I have very little experience with PyBricks and PoweredUp in general, I apologize in advance if I have misunderstood or missed something).

Thx!
Jean

**Comment by laurensvalk** [comment]
Thanks for opening this discussion.

We have a few ideas for generic communication between hubs and other peripherals: #262.

Scenario 3B seems to be the closest to what you might need. This is a generic serial communication; a dedicated remote command protocol could be added on top.

What do you think? 

There's even an idea for direct control from some gamepads, see scenario 4.

**Reply by JeanRev** [reply]
Thanks for the linked ticket @laurensvalk , it looks exciting! Yes, I think we can move the discussion over there. I'll make a comment with my thoughts. Thanks again
